lib/phlex/sgml.rb

```diff
@@ -433,5 +449,78 @@ def __attributes__(attributes, buffer = +"")
 
 			buffer
 		end
+
+		# @api private
+		def __classes__(c)
+			case c
+			when String
+				c
+			when Symbol
+				c.name
+			when Array, Set
+				c.filter_map { |c| __classes__(c) }.join(" ")
+			when Hash
+				c.filter_map { |c, add|
+					next unless add
+					case c
+						when String then c
+						when Symbol then c.name.tr("_", "-").delete_suffix("?")
+						else raise ArgumentError, "Class keys should be Strings or Symbols."
+					end
+				}.join(" ")
+			when nil, false
+				nil
+			else
+				if c.respond_to?(:to_phlex_attribute_value)
+					c.to_phlex_attribute_value
+				elsif c.respond_to?(:to_str)
+					c.to_str
+				else
+					c.to_s
+				end
+			end
+		end
+
+		# @api private
+		def __styles__(s)
+			style = case s
+			when String
+				s
+			when Symbol
+				s.name
+			when Integer, Float
+				s.to_s
+			when Array, Set
+				s.filter_map { |s| __styles__(s) }.join
+			when Hash
+				buffer = +""
+				s.each do |k, v|
+					prop = case k
+						when String then k
+						when Symbol then k.name.tr("_", "-")
+						else raise ArgumentError, "Style keys should be Strings or Symbols."
+					end
+
+					value = __styles__(v)
+
+					if value
+						buffer << prop << ":" << value
+					end
+				end
+				buffer
+			when nil, false
+				return nil
+			else
+				if s.respond_to?(:to_phlex_attribute_value)
+					s.to_phlex_attribute_value
+				elsif s.respond_to?(:to_str)
+					s.to_str
+				else
+					s.to_s
+				end
+			end
+
+			style[-1] == ";" ? style : style + ";"
```

Comment:
Two things here:

1. Is it important that style strings end with a semicolon? It looks like they work either way.
2. We can save an allocation and get a speed up by changing this check to use `end_with?` instead of comparing against a slice of the string.

Reply:
1. When combining a user-provided string in the Array and Hash variants, joining with `;` might result in `;;` if the user already had that suffix. While that's still valid CSS, it felt nicer for `__style__` output to be consistent
2. 👍 on using `end_with?` if it's faster